### PR TITLE
more signature allocation optimizations

### PR DESCRIPTION
**Pull request by froydnj**
cc @zanker-stripe 

### Motivation

Follow on from #6991 with suggestions from the team at Stripe.  We can't modify the `parameters` array in-place in `new_untyped`, unfortunately, because it's coming from Ruby itself, which freezes the array.

I was too lazy to merge the loops in `new_untyped`.  Maybe another day.

This is probably going to conflict with #6993 cc @maruth-stripe .

### Test plan

See included automated tests.

### CODE REVIEW

File: gems/sorbet-runtime/lib/types/private/methods/signature.rb [hunk 1]
```diff
@@ -82,8 +83,13 @@ def initialize(method:, method_name:, raw_arg_types:, raw_return_type:, bind:, m
     if parameters.size != raw_arg_types.size
       raise "The declaration for `#{method.name}` has arguments with duplicate names"
     end
+    params_size = parameters.size
+    i = 0
+    while i < params_size
```
Comment (jez) [inline gems/sorbet-runtime/lib/types/private/methods/signature.rb:88]: This is a completely minor thing but I always liked the pattern of structuring this like
```suggestion
    i = -1
    while i < params_size
      i += 1
```

So that all the index logic happens in one place.

**Review by jez — APPROVED** [review]

File: gems/sorbet-runtime/lib/types/private/methods/signature.rb (hunk 1 above)
Comment (froydnj) [reply gems/sorbet-runtime/lib/types/private/methods/signature.rb:88]: The proposed change is not the common pattern for sorbet-runtime, and it also has an edge case when `params_size` is 0.  I'm opting to leave this as is, though I do agree that for this particular function, having some 50 lines between test and increment is suboptimal.

Comment (jez) [reply gems/sorbet-runtime/lib/types/private/methods/signature.rb:88]: Oh right I forgot that i usually use this with for auto loops
